playbooks/setup_openstack_instances.yml

```diff
@@ -58,16 +76,74 @@
       loop_control:
         loop_var: instance
 
-    - name: Lock instances to prevent changes
+    - name: Create data volumes
+      os_volume:
+        display_name: "{{ item[1]['name'] }}"
+        volume_type: "{{ item[1]['type'] }}"
+        size: "{{ item[1]['size'] }}"
+        state: present
+        cloud: "{{ cloud_name | default('public_cloud') }}"
+        region_name: "{{ item[0]['region'] }}"
+        wait: yes
+        timeout: 900
+      when:
+        - "{{ item[0]['state'] | default('present') != 'absent' }}"
+      with_subelements:
+        - "{{ instance_list | default([]) }}"
+        - volumes
+        - skip_missing: True
+
+    - name: Delete data volumes on deleted instances
+      os_volume:
+        display_name: "{{ item[1]['name'] }}"
+        volume_type: "{{ item[1]['type'] }}"
+        size: "{{ item[1]['size'] }}"
+        state: absent
+        cloud: "{{ cloud_name | default('public_cloud') }}"
+        region_name: "{{ item[0]['region'] }}"
+        wait: yes
+        timeout: 900
+      when:
+        - "{{ item[0]['state'] | default('present') == 'absent' and item[1]['state'] | default('present') == 'absent' }}"
+      with_subelements:
+        - "{{ instance_list | default([]) }}"
+        - volumes
+        - skip_missing: True
+
+    # NOTE(mattt):
+    # At present we do not support detaching / deleting volumes attached to a
+    # running instance. We'd need to do some poking around inside the instance
+    # to check if the disk is in use before we know it's safe to detach /
+    # delete, and this will require a bunch more testing.
+    - name: Attach the data volumes
+      os_server_volume:
+        server: "{{ item[0]['name'] }}"
+        volume: "{{ item[1]['name'] }}"
+        device: "{{ item[1]['device'] }}"
+        state: present
+        cloud: "{{ cloud_name | default('public_cloud') }}"
+        region_name: "{{ item[0]['region'] }}"
+        wait: yes
+        timeout: 900
+      when:
+        - "{{ item[0]['state'] | default('present') != 'absent' and item[1]['state'] | default('present') != 'absent' }}"
+      with_subelements:
+        - "{{ instance_list | default([]) }}"
+        - volumes
+        - skip_missing: True
+
```

Comment:
Why not doing all those operations based on present / absent state on a higher level, in a include for example, and simplify the conditions?

Alternatively, instead of doing it in a serial way, you could do it in parallel, by generating a list of hostnames based on instance_list, and then use a connection local to do your openstack module things. That would be faster for many elements.

Reply:
I'll ping you tomorrow @evrardjp to discuss this point a little better.  I'd like the logic to be clearer if possible.